database.py

```diff
@@ -0,0 +1,135 @@
+import config
+import utils
+
+import MySQLdb
+
+import contextlib
+import os
+import time
+import warnings
+
+REPLICA_MY_CNF = os.getenv(
+    'REPLICA_MY_CNF', os.path.expanduser('~/replica.my.cnf'))
+TOOLS_LABS_CH_MYSQL_HOST = 'tools.db.svc.eqiad.wmflabs'
+
+class _RetryingConnection(object):
+    '''
+    Wraps a MySQLdb connection, handling retries as needed.
+    '''
+
+    def __init__(self, connect, sleep = time.sleep):
+        self._connect = connect
+        self._sleep = sleep
+        self._do_connect()
+
+    def _do_connect(self):
+        self.conn = self._connect()
+        self.conn.ping(True) # set the reconnect flag
+
+    def execute_with_retry(self, operations, *args, **kwds):
+        max_retries = 5
+        for retry in range(max_retries):
+            try:
+                with self.conn.cursor() as cursor:
+                    return operations(cursor, *args, **kwds)
+            except MySQLdb.OperationalError:
+                if retry == max_retries - 1:
+                    raise
+                else:
+                    self._sleep(2 ** retry)
+                    self._do_connect()
+            else:
+                break
+
+    def execute_with_retry_s(self, sql, *args):
+        def operations(cursor, sql, *args):
+            cursor.execute(sql, args)
+            if cursor.rowcount > 0:
+                return cursor.fetchall()
+            return None
+        return self.execute_with_retry(operations, sql, *args)
+
+    def __getattr__(self, name):
+        return getattr(self.conn, name)
+
+@contextlib.contextmanager
+def ignore_warnings():
+    warnings.filterwarnings('ignore', category = MySQLdb.Warning)
+    yield
+    warnings.resetwarnings()
+
+def _connect(**kwds):
+    return MySQLdb.connect(charset = 'utf8mb4', autocommit = True, **kwds) 
+
+def _connect_to_cd_mysql():
+    kwds = {'read_default_file': REPLICA_MY_CNF}
+    if utils.running_in_tools_labs():
+        kwds['host'] = TOOLS_LABS_CH_MYSQL_HOST
+    return _connect(**kwds)
+
+def _connect_to_wp_mysql(cfg):
+    kwds = {'read_default_file': REPLICA_MY_CNF}
+    if utils.running_in_tools_labs():
+        # Get the project database name (and ultimately the database server's
+        # hostname) from the name of the database we want, as per:
+        # https://wikitech.wikimedia.org/wiki/Help:Tool_Labs/Database#Naming_conventions
+        xxwiki = cfg.database.replace('_p', '')
+        kwds['host'] = '%s.analytics.db.svc.eqiad.wmflabs' % xxwiki
+    return _connect(**kwds)
+
+def _make_tools_labs_dbname(cursor, database, lang_code):
+    cursor.execute("SELECT SUBSTRING_INDEX(USER(), '@', 1)")
+    user = cursor.fetchone()[0]
+    return '%s__%s_%s' % (user, database, lang_code)
+
+def _use(cursor, database, lang_code):
+    cursor.execute('USE %s' % _make_tools_labs_dbname(
+        cursor, database, lang_code))
+
+def init_scratch_db():
```

Comment:
You'll also want the function for installing the scratch database, right? https://github.com/eggpi/citationhunt/blob/master/chdb.py#L241

Basically in CH we create a "scratch" database, then atomically swap the real one with it, at which point it becomes the new real one that is used for serving.

Reply:
Oh yes, I need the function as well.